### ISSUE TITLE
Sidekiq Enterprise Periodic - duplicated jobs

### ISSUE THREAD

**Issue by camol** [mperham/sidekiq#3773]
Ruby version: 2.4.2
Sidekiq / Pro / Enterprise version(s):
5.0.5 / 3.7.1 / 1.6.1

Initializer:
```
require 'sidekiq'

if ENV['SIDEKIQ_REDIS_URL']
  Sidekiq.configure_server do |config|
    config.super_fetch!
    config.reliable_scheduler!
    config.failures_default_mode = :all

    config.redis = { url: ENV.fetch('SIDEKIQ_REDIS_URL') }
    config.periodic do |mgr|
      mgr.register '5 0 * * *', Stuff::SomeJob
    end
  end

  Sidekiq.configure_client do |config|
    config.redis = { url: ENV.fetch('SIDEKIQ_REDIS_URL') }
  end
end

Sidekiq::Enterprise.unique! unless Rails.env.test?

Sidekiq.default_worker_options["retry"] = 3
Sidekiq.default_worker_options["unique_for"] = 10.seconds

Sidekiq::Enterprise::Crypto.enable(active_version: 1) do |version|
  File.open("./config/crypto/secret.#{Rails.env}.#{version}.key").read
end
```

Hi, we are using Sidekiq enterprise with periodic for running CRON jobs. We lately spotted that all of our CRON jobs are scheduled twice even though the [WIKI](https://github.com/mperham/sidekiq/wiki/Ent-Periodic-Jobs) says that `Sidekiq will ensure that only a single process creates jobs so if you are running 1 or 100 processes, you don't need to worry about duplicate job creation.`. We can clearly see it in our logs:

```
00:05:01.557 ip-10-1-2-147 /var/log/eb-docker/containers/eb-current-app/0bb79a17029b-stdouterr.log 00:05:00 worker.1            | 2018-02-23T00:05:00.015Z 20 TID-13tnpo INFO: Enqueued periodic job Stuff::SomeJob with JID d423f17bc101c07061df4af3 for 2018-02-24 00:05:00 +1100
00:05:02.143 ip-10-100-2-113 /var/log/eb-docker/containers/eb-current-app/378ddfb1ad3e-stdouterr.log 00:05:00 workerspot.1        | 2018-02-23T00:05:00.007Z 22 TID-ov0c98g50 INFO: Enqueued periodic job Stuff::SomeJob with JID eb76aa41a1a94b03572116aa for 2018-02-24 00:05:00 +1100
```

each job even gets its own JID. I'm not able to reproduce it locally but I see that the leader is always doing the CRON thing. We run multiple sidekiq processes and the amount of them is dynamic since we scale our system (we use docker and we run single sidekiq process with different number of threads on separate docker containers). The difference in processing time for them is only about few milliseconds.

**Comment by mperham** [comment]
Sounds like you have two leaders.  Do you have a blue/green deploy strategy?  What is SIDEKIQ_REDIS_URL?

**Comment by camol** [comment]
`SIDEKIQ_REDIS_URL` is the environment variable which defines the url for our redis. It is the same for every instance of the sidekiq process. So every sidekiq process is connected to the same redis. We use AWS elasticbeanstalk environment and docker. Right I can see that we have only one leader (on the web-ui), the leadership is something which is done out of our control. Our servers scale up and down depending on the traffic or planned heavy overnight processing.

**Comment by mperham** [comment]
Is it possible you have a network partition so that the two leaders are talking to two different Redises, e.g. in a failover scenario? Maybe the two leaders are in different AZs? There can only be one leader per Redis but if they somehow have different config or failover...

**Comment by camol** [comment]
Every sidekiq process shares exactly same config. So I'm not sure how there can be two different leaders. It happens for every periodic job no matter when the job is fired by periodic (even during the day when we do not scale, so there is no option for leadership being bot settled already).

**Comment by mperham** [comment]
Are you using Redis Cluster, Redis Sentinel, any Redis failover?  Give me more hard data to diagnose or I can't help.

**Comment by mperham** [comment]
If you can get me any logging from those two processes that isn't normal job processing output, that might help.

**Comment by camol** [comment]
We use AWS ElasticCache Redis cluster, with primary and slave node with [Multi A-Z](https://docs.aws.amazon.com/AmazonElastiCache/latest/UserGuide/AutoFailover.html) enabled. In our app the `SIDEKIQ_REDIS_URL` points to the primary node. I'm trying to find anything useful but in our aggregation logger I can not find nothing more than I posted in the initial issue comment. Each process just does the periodic thing and that's it. Does anything more comes to you mind which I can check to provide some more useful insight?

**Comment by mperham** [comment]
I think there’s your problem. Split brain / network partition across AZs.  Turn that off.

**Comment by camol** [comment]
I will try that, but this will also mean that I won't have the replication and failover mechanism. So when somehow my primary node will not be available (e.g. no free connections) the processing will fail because of redis timeout I believe.

**Comment by mperham** [comment]
I think multi-az is the main issue. You can disable that at first. It’ll still get you a primary and replica in one AZ.

**Comment by camol** [comment]
Ok thx will let you know about the results.

**Comment by camol** [comment]
The other thing which is interesting is that the "CRON" tab in web-ui for these jobs in the "History" section at the bottom shows that the job was processed only once where the logs show differently. So the web-ui looks like everything is totally fine :\.

**Comment by serek** [comment]
@mperham any comment on the `web-ui` showing single entry rather than two?

**Comment by camol** [comment]
@mperham we turned off the Multi A-Z feature but we still observed duplicated periodic actions. We will try with removing the replica as well and leave the primary only and will if it helps finally.

**Comment by camol** [comment]
Unfortunately this also did not help. Is there a way that I would be able to somehow programmatically double check and confirm that we actually have these two leaders even if the logs messages says so?

**Comment by mperham** [comment]
Look for leader in the ps output:

> $ ps aux | grep leader
> mikeperham       98271   0.0  0.6  2577072  53084 s001  S+    3:15PM   0:01.62 sidekiq 5.1.2 myapp [0 of 25 busy] **leader**

**Comment by camol** [comment]
I've checked both instances which in our logs are performing as the leaders and only one of them has the `leader` label on the process list.

**Comment by mperham** [comment]
Once a Sidekiq process is no longer leader, it should not enqueue periodic jobs.  Are you saying that the logs show that a non-leader process is still enqueueing periodic jobs now?  Reviewing the code, I can't see a way that is possible at first glance but I'm fallible.

It would help to review any "unusual" backtraces in your log if that is easy to obtain.

**Comment by camol** [comment]
Yes unfortunately that's the case. Somehow the non-leader process is also scheduling a periodic job and the common thing that I see in the logs is that it is being done always after the leader. We use the log aggregator and it only shows the info that a periodic job was picked up and some other processing of the actual jobs. There is nothing there apart from that which could be unusual.

Is it somehow possible that the process actually cached the info about its previous leadership (assuming that it was a "proper" leader at some point) and is acting based on it despite that it is not being shown as the label next to the process on the list?

**Comment by mperham** [comment]
The leadership token is a timestamp.  After 1 minute, it's invalid and the process will no longer consider itself a leader unless it can renew the token in Redis. I can't explain how this is happening but I assume you have restarted the process in question?

Is it possible you have logs from another environment or production copy?

**Comment by camol** [comment]
Can I somehow try to find that leadership token in redis directly?

**Comment by mperham** [comment]
```
$ redis-cli get dear-leader
"MikeBookPro.local:13219:fa062e7249e2"
```

**Comment by mperham** [comment]
The `dear-leader` key has a 1 minute TTL in Redis.

**Comment by camol** [comment]
I see only one entry in redis so all looks fine on this level. I killed the process and boot it up again but still acts as "fake leader". I'm going to set the sidekiq logger level to debug and maybe I will be able to get back with more data to analyze.

**Comment by mperham** [comment]
Does SIDEKIQ_REDIS_URL resolve to the same host:port on both instances?

**Comment by camol** [comment]
Yup it is

**Comment by mperham** [comment]
I'm frustrated.  At this point I have no idea what the problem might be.  The cron feature has had no bugs reported for it in the last 6-12 months; it's been quite stable so I have to think that something is really out of whack with the environment.  Stick with a simple primary/replica Redis setup without clustering or multi-AZ.  Beyond that, you'll need to do some more debugging to figure out why a non-leader process still thinks it can create periodic jobs.

**Comment by camol** [comment]
I found another strange thing about the non-leader. I saw that the process picks up the jobs even from the queue which are not defined for him. It seems to finish them usually very quickly e.g. 0.083 sec when the proper workers perform them in more reasonable time.

**Comment by camol** [comment]
We finally moved to system CRON for this. But I managed based on the logs to see that the jobs which are being picked up from the queue (checked for the ones put by the CRON in) are showing similar symptoms that the process which was previously the "fake leader" also picks up these jobs from the queue and leaves them right a way somehow. This is always one and the same process from the particular instance.
